[Bug] Fix bug that delete stmt with filter condition delete all data from table on segment v2

caiconghui opened the pull request:
Fix #3942

Commit:
fix

Commit:
Fix bug that delete stmt with filter condition delete all data from table on segment v2
   When we get different columns's row ranges by column_delete_conditions, we should use union operation instead of intersection operation to get final get final row ranges.

chaoyli commented:
You can describe the solution in the commit message.

caiconghui commented:
> You can describe the solution in the commit message.

done

morningman commented:
It is strange that the case you described in issue is truly has problem. But the following case is OK:

```
mysql> create table tbl4 (k1 int, k2 int) distributed by hash(k1) buckets 1 properties("replication_num" = "1");
Query OK, 0 rows affected (0.01 sec)

mysql> insert into tbl4 values(1,2),(3,4);
Query OK, 2 rows affected (0.04 sec)
{'label':'insert_20ddaec3-5d29-43a9-b065-7d3d680e9d33', 'status':'VISIBLE', 'txnId':'3014'}

mysql> select * from tbl4;
+------+------+
| k1   | k2   |
+------+------+
|    1 |    2 |
|    3 |    4 |
+------+------+
2 rows in set (0.01 sec)

mysql> delete from tbl4 where k2=2 and k1=1;
Query OK, 0 rows affected (0.04 sec)
{'label':'delete_1e19b1ba-0a42-48ee-989c-813758ea9a36', 'status':'VISIBLE', 'txnId':'3015'}

mysql> select * from tbl4;
+------+------+
| k1   | k2   |
+------+------+
|    3 |    4 |
+------+------+
1 row in set (0.01 sec)
```

morningman commented:
I think the root cause is we lost the relationship of the 2 delete condition in same delete stmt.
For example, the following 2 cases should has different result:

Base data:
```
k1, k2
1, 2
1, 3
```

case 1:
`delete from tbl where k1=1 and k2=2`;
case 2:
`delete from tbl where k1=1;`
`delete from tbl where k2=2;`

We treat the above 2 cases as same, which is incorrect.

So I think only change the `intersect` to `union` is not the right way to fix the bug.

morningman commented:
I am not familiar with this part. But following is the  pseudocode I thought:

```
All DeleteConditions:

DeleteConditions 1: k1 = 1 and k2 = 2;
DeleteConditions 2: k1 = 4
DeleteConditions 3: k1 = 5


RowRanges result;
foreach(DeleteConditions dcs : All DeleteConditions) {
    RowRanges rr;
    foreach(DeleteCondition dc : dcs) {
        rr = rr UNION (rows filtered by dc)
    }
    result = result INTERSECT rr;
}
return result;


```

caiconghui commented:
@morningman it is a good implementation to filter block which not need read, although  only change the intersect to union can fix the bug but not an efficient fix.

caiconghui committed:
Merge branch 'bug_delete' of https://github.com/caiconghui/incubator-doris into bug_delete

caiconghui committed:
[Bug] Fix bug that delete stmt with filter condition delete all data from table on segment v2

The root cause is that we lost the relationship of the two delete conditions in same delete stmt.
Base data:
    k1, k2
    1,  2
    1,  3
case 1:
   delete from tbl where k1=1 and k2=2;
case 2:
   delete from tbl where k1=1;
   delete from tbl where k2=2;

We treat the above 2 cases as same, which is incorrect.
So we need to process every rowset of delete conditions separately.

caiconghui committed:
Revert commit for log to remove unused log

This reverts commit e56dabb1a1ceb0c1120f9cd3d260f58c741562fb.

caiconghui committed:
Add docs/contents and be/build_Release to .gitignore file

morningman commented on be/src/olap/rowset/segment_v2/segment_iterator.cpp:
```diff
@@ -253,30 +253,37 @@ Status SegmentIterator::_get_row_ranges_from_conditions(RowRanges* condition_row
+            RowRanges single_delete_condition_row_ranges = RowRanges::create_single(num_rows());
+            RETURN_IF_ERROR(
+                    _column_iterators[cid]->get_row_ranges_by_zone_map(
+                            column_cond,
```
Should the `column_cond` be null here?
We already filter the data by column conditions before.

caiconghui replied:
@morningman here we use column_cond to skip zone that not need read, so we don't need to process skip zone with delete conditions

morningman replied:
OK

morningman commented on be/src/olap/rowset/segment_v2/segment_iterator.cpp:
```diff
@@ -253,30 +253,37 @@ Status SegmentIterator::_get_row_ranges_from_conditions(RowRanges* condition_row
-        if (_opts.conditions != nullptr) {
-            column_cond = _opts.conditions->get_column(cid);
-        }
+        CondColumn* column_cond = _opts.conditions->get_column(cid);
```
`_opts.conditions` may be `nullptr` here.
You need to check it before use, like the old way.

caiconghui replied:
done

Commit:
fix core caused by null conditions of opts

morningman reviewed (approved):
LGTM
